ShaiRandom/Generators/AbstractRandom.cs

```diff
@@ -0,0 +1,560 @@
+        public double NextExclusiveDouble(double outerBound)
+        {
+            return NextExclusiveDouble() * outerBound;
+        }
+
+        /// <inheritdoc />
+        public double NextExclusiveDouble(double innerBound, double outerBound)
+        {
+            return innerBound + NextExclusiveDouble() * (outerBound - innerBound);
+        }
+
+
+        /// <inheritdoc />
+        public float NextExclusiveFloat()
+        {
+            // return ((NextUInt() >> 9) + 1u) * 5.960464E-8f;
+            long bits = NextLong();
+            return BitConverter.Int32BitsToSingle((1089 + (int)(BitConverter.DoubleToInt64Bits(-0x7FFFFFFFFFFFF001L | bits) >> 52) << 23) | ((int)~bits & 0x007FFFFF));
+        }
+
+        /// <inheritdoc />
+        public float NextExclusiveFloat(float outerBound)
+        {
+            return NextExclusiveFloat() * outerBound;
+        }
+
+        /// <inheritdoc />
+        public float NextExclusiveFloat(float innerBound, float outerBound)
+        {
+            return innerBound + NextExclusiveFloat() * (outerBound - innerBound);
+        }
+
+        /// <inheritdoc />
+        public virtual ulong Skip(ulong distance)
+        {
+            throw new NotSupportedException("Skip() is not implemented for this generator.");
+        }
+
+        /// <summary>
+        /// (Optional) If implemented, jumps the generator back to the previous state and returns what NextULong() would have produced at that state.
+        /// </summary>
+        /// <remarks>
+        /// The default implementation calls <see cref="Skip(ulong)"/> with the equivalent of (ulong)(-1L) . If Skip() is not implemented, this throws a NotSupportedException.
+        /// Be aware that if Skip() has a non-constant-time implementation, the default here will generally take the most time possible for that method.
+        /// </remarks>
+        /// <returns>The result of what NextULong() would return at the previous state.</returns>
+        public virtual ulong PreviousULong()
+        {
+            return Skip(0xFFFFFFFFFFFFFFFFUL);
+        }
+
+        /// <inheritdoc />
+        public abstract IEnhancedRandom Copy();
+
+        /// <summary>
+        /// Sets each state in this IEnhancedRandom to the corresponding state in the other IEnhancedRandom.
+        /// This generally only works correctly if both objects have the same class.
+        /// </summary>
+        /// <param name="other">Another IEnhancedRandom that almost always should have the same class as this one.</param>
+        public void SetWith(IEnhancedRandom other)
+        {
+            int myCount = StateCount, otherCount = other.StateCount;
+            int i = 0;
+            for (; i < myCount && i < otherCount; i++)
+            {
+                SetSelectedState(i, other.SelectState(i));
+            }
+            for (; i < myCount; i++)
+            {
+                SetSelectedState(i, 0xFFFFFFFFFFFFFFFFUL);
+            }
+        }
```

Comment:
SetWith() should probably be moved to an extension method to avoid some extra effort while implementing.

Reply:
Moved to extension method.